File: apps/worker/src/camera-policy.ts
/**
 * Single source for canonical camera math shared by the revision engine, the
 * Corona adapter, and the DCC-side camera runners. Pure and DCC-independent:
 * no pymxs/MAXScript concept crosses this boundary.
 *
 * 3ds Max's MAXScript `Camera.fov` (and its trig functions `atan`/`tan`) are
 * degrees-based; `fovRadians` here is a genuine radian value. A DCC-side
 * caller assigning to `camera.fov` must convert with `radiansToDegrees`, and
 * a caller reading `camera.fov` back must convert with `degreesToRadians`
 * before comparing against a canonical radian value (Spike 8H's degrees vs
 * radians defect: `camera.fov = fovRadians` directly silently pointed a
 * temporary camera at ~1.3 degrees instead of the intended ~74).
 */

export type CameraVector3 = readonly [number, number, number];

const CANONICAL_ANGLE_PRECISION = 6;

/**
 * Canonical serialization for a derived camera Euler angle: fixed 6-decimal
 * rounding so two derivations from the same position/target always
 * serialize to the identical JSON number, matching the precision already
 * used by the hand-authored Golden camera fixtures (e.g. rev11's
 * `camera_living_a` rotation). This is a canonical-representation concern
 * only — a live DCC observation is compared with its own, deliberately
 * looser angular tolerance (see `verify_canonical_camera_state.py`'s
 * `ANGLE_TOLERANCE`), which this function must never replace.
 */
export function canonicalCameraAngle(value: number): number {
  const factor = 10 ** CANONICAL_ANGLE_PRECISION;
  return Math.round(value * factor) / factor;
}

export function radiansToDegrees(radians: number): number {
  return (radians * 180) / Math.PI;
}

export function degreesToRadians(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

export function deriveCameraFovRadians(focalLengthMm: number, sensorWidthMm: number): number {
  return 2 * Math.atan(sensorWidthMm / (2 * focalLengthMm));
}

export function deriveCameraFovDegrees(focalLengthMm: number, sensorWidthMm: number): number {
  return radiansToDegrees(deriveCameraFovRadians(focalLengthMm, sensorWidthMm));
}

/**
 * Canonical look-at Euler derivation (degrees), matching the exact
 * atan2-based convention already proven by the Golden fixture chain and the
 * 8E/8H DCC runners' `_look_at_rotation`. `position` and `target` must
 * differ; a coincident position/target has no defined orientation.
 */
/**
 * Rounded to a fixed precision because JS `Math.hypot` and Python's
 * `math.hypot` compute the same distance with different internal scaling
 * algorithms and disagree in the ~13th significant digit — far below any
 * meaningful precision at millimeter scale, but enough to fail an exact
 * cross-process evidence comparison if left at full float precision.
 */
export function targetDistanceMm(position: CameraVector3, target: CameraVector3): number {
  const dx = target[0] - position[0];
  const dy = target[1] - position[1];
  const dz = target[2] - position[2];
  return Math.round(Math.hypot(dx, dy, dz) * 1_000_000) / 1_000_000;
}

export function deriveLookAtRotationEuler(
  position: CameraVector3,
  target: CameraVector3,
): [number, number, number] {
  const dx = target[0] - position[0];
  const dy = target[1] - position[1];
  const dz = target[2] - position[2];
  const horizontal = Math.hypot(dx, dy);
  if (horizontal === 0 && dz === 0) {
    throw new Error("Camera position and target must differ");
  }
  const pitch = radiansToDegrees(Math.atan2(dz, horizontal));
  const yaw = (radiansToDegrees(Math.atan2(dy, dx)) + 270) % 360;
  return [canonicalCameraAngle(pitch), canonicalCameraAngle(0), canonicalCameraAngle(yaw)];
}

File: tools/3ds-max/python/camera_policy.py
"""Single source for canonical camera math shared by the DCC-side mutation
and verification scripts. Pure and DCC-independent (no pymxs import here),
mirroring `apps/worker/src/camera-policy.ts` exactly so the TS and Python
sides can never independently drift.

3ds Max's MAXScript `Camera.fov` (and its trig functions `atan`/`tan`) are
degrees-based; `fov_radians()` here returns a genuine radian value. A caller
assigning to `camera.fov` must convert with `math.degrees(...)`, and a
caller reading `camera.fov` back must convert with `math.radians(...)`
before comparing against a canonical radian value (the Spike 8H degrees vs
radians defect: assigning a radian number directly to `camera.fov` silently
pointed a camera at ~1.3 degrees instead of the intended ~74).
"""

from __future__ import annotations

import math

ROTATION_ANGLE_TOLERANCE = 0.001


def fov_radians(focal_length_mm: float, sensor_width_mm: float) -> float:
    return 2.0 * math.atan(float(sensor_width_mm) / (2.0 * float(focal_length_mm)))


def fov_degrees(focal_length_mm: float, sensor_width_mm: float) -> float:
    return math.degrees(fov_radians(focal_length_mm, sensor_width_mm))


def look_at_rotation_euler(position: list[float], target: list[float]) -> tuple[float, float, float]:
    dx = float(target[0]) - float(position[0])
    dy = float(target[1]) - float(position[1])
    dz = float(target[2]) - float(position[2])
    horizontal = math.hypot(dx, dy)
    if horizontal == 0 and dz == 0:
        raise ValueError("Camera position and target must differ")
    pitch = math.degrees(math.atan2(dz, horizontal))
    yaw = (math.degrees(math.atan2(dy, dx)) + 270.0) % 360.0
    return pitch, 0.0, yaw


def target_distance_mm(position: list[float], target: list[float]) -> float:
    """Rounded to a fixed precision because JS `Math.hypot` and Python's
    `math.hypot` compute the same distance with different internal scaling
    algorithms and disagree in the ~13th significant digit — far below any
    meaningful precision at millimeter scale, but enough to fail an exact
    cross-process evidence comparison if left at full float precision.
    """
    dx = float(target[0]) - float(position[0])
    dy = float(target[1]) - float(position[1])
    dz = float(target[2]) - float(position[2])
    return round(math.hypot(dx, dy, dz), 6)


def implied_target(
    position: list[float], rotation_euler: list[float], target_distance: float
) -> list[float]:
    """Inverse of `look_at_rotation_euler`: reconstructs the physical point a
    camera at `position` with `rotation_euler` is looking at, `target_distance`
    away. Used to derive the OBSERVED target from physical camera state
    (position + orientation + targetDistance) rather than trusting only
    stored target metadata (Spike 8I observed-target requirement).
    """
    pitch_rad = math.radians(float(rotation_euler[0]))
    yaw_rad = math.radians(float(rotation_euler[2]))
    horizontal = float(target_distance) * math.cos(pitch_rad)
    dz = float(target_distance) * math.sin(pitch_rad)
    angle = yaw_rad - math.radians(270.0)
    dx = horizontal * math.cos(angle)
    dy = horizontal * math.sin(angle)
    return [float(position[0]) + dx, float(position[1]) + dy, float(position[2]) + dz]


def close(left: float, right: float, tolerance: float = 0.01) -> bool:
    return abs(float(left) - float(right)) <= tolerance


def angle_close(left: float, right: float, tolerance: float = ROTATION_ANGLE_TOLERANCE) -> bool:
    difference = (float(left) - float(right) + 180.0) % 360.0 - 180.0
    return abs(difference) <= tolerance
